src/p2p/swarm.rs

```diff
@@ -406,27 +431,183 @@ fn connection_point_addr(cp: &ConnectedPoint) -> &Multiaddr {
+                        // it's the sort of difficulty with the cli functionality in general: what
+                        // does it mean to connect to a peer? one way to look at it would be to
+                        // make the peer a "pinned peer" or "friend" and to keep the connection
+                        // alive at all costs. perhaps that is something for the next round.
+                        // another aspect would be to fail this future because there was no
+                        // `inject_connected`, only `inject_connection_established`. taking that
+                        // route would be good; it does however leave the special case of adding
+                        // another connection, which does add even more complexity than it exists
+                        // at the present.
+                        res.unwrap();
+
+                        // just to confirm that there are no connections.
+                        assert_eq!(Vec::<Multiaddr>::new(), swarm1.connections_to(&peer2_id));
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn wrong_peerid() {
+        let (peer1_id, trans) = mk_transport();
+        let mut swarm1 = SwarmBuilder::new(trans, SwarmApi::default(), peer1_id)
+            .executor(Box::new(ThreadLocalTokio))
+            .build();
+
+        let (peer2_id, trans) = mk_transport();
+        let mut swarm2 = SwarmBuilder::new(trans, SwarmApi::default(), peer2_id)
+            .executor(Box::new(ThreadLocalTokio))
+            .build();
+
+        let peer3_id = Keypair::generate_ed25519().public().into_peer_id();
+
+        Swarm::listen_on(&mut swarm1, "/ip4/127.0.0.1/tcp/0".parse().unwrap()).unwrap();
+
+        let address;
+
+        loop {
+            if let SwarmEvent::NewListenAddr(addr) = swarm1.next_event().await {
+                // wonder if there should be a timeout?
+                address = addr;
+                break;
+            }
+        }
+
+        let mut fut = swarm2
+            .connect(
+                MultiaddrWithoutPeerId::try_from(address)
+                    .unwrap()
+                    .with(peer3_id),
+            )
+            .unwrap()
+            // remove the private type wrapper
+            .map_err(|e| e.into_inner());
+
+        loop {
+            tokio::select! {
+                _ = swarm1.next_event() => {},
+                _ = swarm2.next_event() => {},
+                res = &mut fut => {
+                    assert_eq!(res.unwrap_err(), Some("Pending connection: Invalid peer ID.".into()));
+                    return;
+                }
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn racy_connecting_attempts() {
+        let (peer1_id, trans) = mk_transport();
+        let mut swarm1 = SwarmBuilder::new(trans, SwarmApi::default(), peer1_id)
```

Comment:
we might want to introduce some helper function for those `tokio`-based `Swarm` creators

Reply:
I don't think the return type is easy to name because of the many associated types, didn't even try this. Lets see.

Reply:
Yes it's easy, thanks to the transport config being boxed which probably is an overall great thing.